### PR TITLE
Add support for cursor based pagination

### PR DESCRIPTION
**Pull request by kmossco**
Since we added cursor based pagination to several of our API's endpoints this adds support to that pagination. In order to use it we can call `next` in the supported endpoints:

```ruby
# example with Plans
plans = []

response = ChartMogul::Plan.all(per_page: 10)

while response
  plans << response.plans if response.plans.any?
  break unless response.has_more

  response = response.next(per_page: 10)
end
```

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # chartmogul-ruby Change Log
 
+## Version 4.0.0 - August 15 2023
```
Comment (swember) [inline changelog.md:3]: Do we really want it to be `4.0.0`? For me it feels more like `3.4.0`. It also should be updated in [version.rb](https://github.com/chartmogul/chartmogul-ruby/blob/main/lib/chartmogul/version.rb).

**Review by swember — CHANGES_REQUESTED** [review]

File: lib/chartmogul/resource_path.rb [hunk 2]
```diff
@@ -43,5 +46,12 @@ def apply_named_params(params)
         end
       end
     end
+
+    def validate_params!(params)
+      invalid = params.keys.find { |param| INVALID_PARAMS.include?(param) }
```
Comment (swember) [inline lib/chartmogul/resource_path.rb:51]: Might keys contain a string?

File: lib/chartmogul/subscription.rb [hunk 3]
```diff
@@ -54,7 +54,7 @@ class Subscriptions < APIResource
     set_resource_root_key :subscriptions
 
     include Concerns::Entries
-    include Concerns::Pageable2
+    include Concerns::PageableWithCursor
```
Comment (swember) [inline lib/chartmogul/subscription.rb:57]: Why other resources have `next` method defined and this does not?

File: lib/chartmogul/subscription_event.rb [hunk 4]
```diff
@@ -68,10 +68,9 @@ class SubscriptionEvents < APIResource
 
     set_resource_root_key :subscription_events
 
-    writeable_attr :meta
-
     include API::Actions::All
     include Concerns::Entries
+    include Concerns::PageableWithCursor
```
Comment (swember) [inline lib/chartmogul/subscription_event.rb:73]: Same here. Why no `next` method?

File: lib/chartmogul/subscription.rb (hunk 3 above)
Comment (kmossco) [reply lib/chartmogul/subscription.rb:57]: The `next` method was there from previous versions and I didn't want to remove it to keep parity, should we then add it to the other resources?

File: lib/chartmogul/resource_path.rb (hunk 2 above)
Comment (kmossco) [reply lib/chartmogul/resource_path.rb:51]: Faraday treats parameters as symbols as far as I can tell, which is also why we do some wrangling of the params here: https://github.com/chartmogul/chartmogul-ruby/blob/c8aa9f89401203246bfc3559738e4bf46669c6b7/lib/chartmogul/resource_path.rb#L14-L16

**Review by swember — APPROVED** [review]

**Review by daesu — APPROVED** [review]
lgtm

**Comment by kmossco** [comment]
Closing this for now, as the PR should now remove support for cursor pagination, and after we merge https://github.com/chartmogul/chartmogul-ruby/pull/144 we will have a ton of conflicts. Will open a new one very soon with the new logic.

**Pull request closed without merging**